diff --git a/zircon/kernel/lib/userabi/vdso/syscalls-arm64.S b/zircon/kernel/lib/userabi/vdso/syscalls-arm64.S
index 643ab840a0c..6bfbf15b6ba 100644
--- a/zircon/kernel/lib/userabi/vdso/syscalls-arm64.S
+++ b/zircon/kernel/lib/userabi/vdso/syscalls-arm64.S
@@ -24,0 +25,7 @@ syscall_entry_end \name \public
+
+.pushsection .rodata
+.global zcore_syscall_entry
+.hidden zcore_syscall_entry
+zcore_syscall_entry:
+    .quad 0xdeadbeaf
+.popsection
diff --git a/zircon/kernel/lib/userabi/vdso/syscalls-x86.S b/zircon/kernel/lib/userabi/vdso/syscalls-x86.S
index d61aafdcec1..01a13629d8d 100644
--- a/zircon/kernel/lib/userabi/vdso/syscalls-x86.S
+++ b/zircon/kernel/lib/userabi/vdso/syscalls-x86.S
@@ -29,29 +29,3 @@
-    .cfi_same_value %r12
-    .cfi_same_value %r13
-    .if \nargs <= 3
-      zircon_syscall \num, \name, \name
-      ret
-    .elseif \nargs <= 6
-      mov %rcx, %r10  // Argument 4
-      zircon_syscall \num, \name, \name
-      ret
-    .elseif \nargs == 7
-      push.spill %r12
-      mov      0x10(%rsp), %r12  // Argument 7
-      mov      %rcx, %r10  // Argument 4
-      zircon_syscall \num, \name, \name
-      pop.reload  %r12
-      ret
-    .elseif \nargs == 8
-      push.spill %r12
-      push.spill %r13
-      mov      0x18(%rsp), %r12  // Argument 7
-      mov      0x20(%rsp), %r13  // Argument 8
-      mov      %rcx, %r10  // Argument 4
-      zircon_syscall \num, \name, \name
-      pop.reload  %r13
-      pop.reload  %r12
-      ret
-    .else
-      .error "maximum 8 syscall arguments"
-    .endif
+    mov $\num, %eax
+    jmpq *zcore_syscall_entry(%rip)
+    .label CODE_SYSRET_\name\()_VIA_\name, global
@@ -61,0 +36,7 @@
+
+.pushsection .rodata
+.global zcore_syscall_entry
+.hidden zcore_syscall_entry
+zcore_syscall_entry:
+    .quad 0xdeadbeaf
+.popsection
diff --git a/zircon/kernel/lib/userabi/vdso/zircon-syscall-arm64.S b/zircon/kernel/lib/userabi/vdso/zircon-syscall-arm64.S
index f36502fc070..e70ab49291c 100644
--- a/zircon/kernel/lib/userabi/vdso/zircon-syscall-arm64.S
+++ b/zircon/kernel/lib/userabi/vdso/zircon-syscall-arm64.S
@@ -10 +10,3 @@
-  svc #0x0
+  adrp x9, zcore_syscall_entry
+  ldr x9, [x9, :lo12:zcore_syscall_entry]
+  br x9
diff --git a/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-arm64.S b/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-arm64.S
index c84f321bdb0..6b57fba6df4 100644
--- a/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-arm64.S
+++ b/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-arm64.S
@@ -13,21 +13 @@ syscall_entry_begin zx_futex_wake_handle_close_thread_exit
-
-  // Store the value into the futex, which should signal that the stack is no longer in use.
-  // atomic_store_explicit(value_ptr, new_value, memory_order_release)
-  stlr w2, [x0]
-
-  // Now the stack might be gone and we can never return!
-
-  // Save the handle argument in a callee-saves register (x19).
-  // Since this function cannot return, do not bother preserving the old contents.
-  mov w19, w3
-
-  zircon_syscall ZX_SYS_futex_wake, zx_futex_wake, zx_futex_wake_handle_close_thread_exit
-  cbnz x0, .Lfutex_wake_fail
-
-  cbz w19, .Lthread_exit
-  mov w0, w19
-  zircon_syscall ZX_SYS_handle_close, zx_handle_close, zx_futex_wake_handle_close_thread_exit
-  cbnz x0, .Lhandle_close_fail
-
-.Lthread_exit:
-  zircon_syscall ZX_SYS_thread_exit, zx_thread_exit, zx_futex_wake_handle_close_thread_exit
+  zircon_syscall 1005, zx_futex_wake_handle_close_thread_exit, zx_futex_wake_handle_close_thread_exit
@@ -39,6 +18,0 @@ syscall_entry_begin zx_futex_wake_handle_close_thread_exit
-  // We can't recover in these cases, since our stack may have been freed.
-.Lfutex_wake_fail:
-  brk #1000
-.Lhandle_close_fail:
-  brk #1000
-
diff --git a/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-x86.S b/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-x86.S
index c5f7af16a24..33f12dfb87e 100644
--- a/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-x86.S
+++ b/zircon/kernel/lib/userabi/vdso/zx_futex_wake_handle_close_thread_exit-x86.S
@@ -15,26 +15,2 @@ syscall_entry_begin zx_futex_wake_handle_close_thread_exit
-
-  // Store the value into the futex, which should signal that the stack is no
-  // longer in use.
-  // atomic_store_explicit(value_ptr, new_value, memory_order_release)
-  mov %edx, (%rdi)
-
-  // Now the stack might be gone and we can never return!
-
-  // Save the handle argument in a callee-saves register (%rbx).  Since this
-  // function cannot return, do not bother preserving the old contents.
-  mov %ecx, %ebx
-
-  zircon_syscall ZX_SYS_futex_wake, zx_futex_wake, zx_futex_wake_handle_close_thread_exit
-  test %eax, %eax
-  jnz .Lfutex_wake_fail
-
-  test %ebx, %ebx
-  jz .Lthread_exit
-
-  mov %ebx, %edi
-  zircon_syscall ZX_SYS_handle_close, zx_handle_close, zx_futex_wake_handle_close_thread_exit
-  test %eax, %eax
-  jnz .Lhandle_close_fail
-
-.Lthread_exit:
-  zircon_syscall ZX_SYS_thread_exit, zx_thread_exit, zx_futex_wake_handle_close_thread_exit
+  mov $1005, %eax
+  jmpq *zcore_syscall_entry(%rip)
@@ -46,6 +21,0 @@ syscall_entry_begin zx_futex_wake_handle_close_thread_exit
-  // We can't recover in these cases, since our stack may have been freed.
-.Lfutex_wake_fail:
-  ud2
-.Lhandle_close_fail:
-  ud2
-
diff --git a/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-arm64.S b/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-arm64.S
index 4d129178c39..f8a6dc7193a 100644
--- a/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-arm64.S
+++ b/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-arm64.S
@@ -15,18 +15 @@ syscall_entry_begin zx_vmar_unmap_handle_close_thread_exit
-
-  // Save the handle argument in a callee-saves register (x19).  Callee-save
-  // that register so we can unwind in the error case.  We only need to save
-  // x19, but we have to keep SP aligned to 16 bytes, so we always push and pop
-  // registers in pairs.
-  stp.spill x19, x20
-  mov x19, x3
-
-  zircon_syscall ZX_SYS_vmar_unmap, zx_vmar_unmap, zx_vmar_unmap_handle_close_thread_exit
-  cbnz x0, .Lvmar_unmap_fail
-
-  // Now the stack is gone and we can never return!
-
-  mov x0, x19
-  zircon_syscall ZX_SYS_handle_close, zx_handle_close, zx_vmar_unmap_handle_close_thread_exit
-  cbnz x0, .Lhandle_close_fail
-
-  zircon_syscall ZX_SYS_thread_exit, zx_thread_exit, zx_vmar_unmap_handle_close_thread_exit
+  zircon_syscall 1004, zx_vmar_unmap_handle_close_thread_exit, zx_vmar_unmap_handle_close_thread_exit
@@ -38,7 +20,0 @@ syscall_entry_begin zx_vmar_unmap_handle_close_thread_exit
-.Lvmar_unmap_fail:
-  ldp.reload x19, x20
-  ret
-
-.Lhandle_close_fail:
-  brk #1000
-
diff --git a/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-x86.S b/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-x86.S
index 01c490a012d..2cdaee0d4ff 100644
--- a/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-x86.S
+++ b/zircon/kernel/lib/userabi/vdso/zx_vmar_unmap_handle_close_thread_exit-x86.S
@@ -13,18 +13,2 @@ syscall_entry_begin zx_vmar_unmap_handle_close_thread_exit
-
-  // Save the handle argument in a callee-saves register (%rbx).
-  // Callee-save that register so we can unwind in the error case.
-  push.spill %rbx
-  mov %ecx, %ebx
-
-  zircon_syscall ZX_SYS_vmar_unmap, zx_vmar_unmap, zx_vmar_unmap_handle_close_thread_exit
-  test %eax, %eax
-  jnz .Lvmar_unmap_fail
-
-  // Now the stack is gone and we can never return!
-
-  mov %ebx, %edi
-  zircon_syscall ZX_SYS_handle_close, zx_handle_close, zx_vmar_unmap_handle_close_thread_exit
-  test %eax, %eax
-  jnz .Lhandle_close_fail
-
-  zircon_syscall ZX_SYS_thread_exit, zx_thread_exit, zx_vmar_unmap_handle_close_thread_exit
+  mov $1004, %eax
+  jmpq *zcore_syscall_entry(%rip)
@@ -36,7 +19,0 @@ syscall_entry_begin zx_vmar_unmap_handle_close_thread_exit
-.Lvmar_unmap_fail:
-  pop.reload %rbx
-  ret
-
-.Lhandle_close_fail:
-  ud2
-
